File: solutions/cpp/Flatten Binary Tree to Linked List.cpp
#include <iostream>
#include <stack>
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    void flatten(TreeNode *root)
    {
        std::stack<TreeNode *> stack;
        while (root)
        {
            if (root->right)
            {
                stack.push(root->right);
            }
            if (root->left)
            {
                stack.push(root->left);
            }
            root->left = NULL;
            TreeNode *temp;
            if (!stack.empty())
            {
                temp = stack.top();
                stack.pop();
                root->right = temp;
                root = temp;
            }
            else
            {
                root->right = NULL;
                root = NULL;
            }
        }
    }
};

File: solutions/python/Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        '''
        #Recursively
        if root:
            temp_left=root.left
            temp_right=root.right
            root.left=None
            self.flatten(temp_left)
            self.flatten(temp_right)
            root.right=temp_left
            current=root
            while(current.right):
                current=current.right
            current.right=temp_right
            '''
        stack=[]
        while root:
            if root.right:
                stack.append(root.right)
            if root.left:
                stack.append(root.left)
            root.left=None
            temp=stack.pop() if len(stack)>0 else None
            root.right=temp
            root=temp